text/0000-futures.md

```diff
@@ -0,0 +1,520 @@
+
+- The most recent revisions were in August, and involved [some insights](https://boats.gitlab.io/blog/post/rethinking-pin/) into how to make the `Pin` APIs even cleaner. These APIs have been [proposed for stabilization](https://github.com/rust-lang/rust/issues/55766), as has [their use as `self` types](https://github.com/rust-lang/rust/issues/55786).
+
+- There are multiple compatibility layers available for using futures 0.1 and 0.3 simultaneously. That's important, because it allows for *incremental* migration of existing production code.
+
+Since the initial futures 0.3 release, relatively little has changed about the core `Future` trait and task system, other than the refinements mentioned above. The actual `Future` trait has stayed essentially as it was back in April.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Future` trait represents an *asynchronous* and lazy computation that may
+eventually produce a final value, but doesn't have to block the current thread
+to do so.
+
+Futures can be constructed through `async` blocks or `async` functions, e.g.,
+
+```rust
+async fn read_frame(socket: &TcpStream) -> Result<Frame, io::Error> { ... }
+```
+
+This `async` function, when invoked, produces a future that represents the
+completion of reading a frame from the given socket. The function signature
+is equivalent to:
+
+```rust
+fn read_frame<'sock>(socket: &'sock TcpStream)
+    -> impl Future<Output = Result<Frame, io::Error>> + 'sock;
+```
+
+Other async functions can *await* this future; see the [companion
+RFC] for full details.
+
+In addition to `async fn` definitions, futures can be built using adapters, much
+like with `Iterator`s. Initially these adapters will be provided entirely "out
+of tree", but eventually they will make their way into the standard library.
+
+Ultimately asynchronous computations are executed by *tasks*, which are
+lightweight threads. In particular, an *executor* is able to "spawn" a
+`()`-producing `Future` as an independent task; these tasks are then
+cooperatively scheduled onto one or more operating system threads. The
+`Executor` trait defines this interface, and the `task` module provides a host
+of related definitions needed when manually implementing `Future`s or
+executors.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core::task` module
+
+The fundamental mechanism for asynchronous computation in Rust is *tasks*, which
+are lightweight threads of execution; many tasks can be cooperatively scheduled
+onto a single operating system thread.
+
+To perform this cooperative scheduling we use a technique sometimes referred to
+as a "trampoline". When a task would otherwise need to block waiting for some
```

Comment:
Is this the same "trampoline" concept which is used in the context of avoiding stack overflows for recursive calls?

Reply:
Yep!